src/System.Runtime.Serialization.Formatters/tests/BinaryFormatterTests.cs

```diff
@@ -0,0 +1,598 @@
+            f.FilterLevel = TypeFilterLevel.Low;
+            Assert.Equal(TypeFilterLevel.Low, f.FilterLevel);
+
+            Assert.Equal(FormatterTypeStyle.TypesAlways, f.TypeFormat);
+            f.TypeFormat = FormatterTypeStyle.XsdString;
+            Assert.Equal(FormatterTypeStyle.XsdString, f.TypeFormat);
+        }
+
+        [Fact]
+        public void SerializeDeserialize_InvalidArguments_ThrowsException()
+        {
+            var f = new BinaryFormatter();
+            Assert.Throws<ArgumentNullException>("serializationStream", () => f.Serialize(null, new object()));
+            Assert.Throws<ArgumentNullException>("serializationStream", () => f.Deserialize(null));
+            Assert.Throws<SerializationException>(() => f.Deserialize(new MemoryStream())); // seekable, 0-length
+        }
+
+        [Theory]
+        [InlineData(FormatterAssemblyStyle.Simple, false)]
+        [InlineData(FormatterAssemblyStyle.Full, true)]
+        public void MissingField_FailsWithAppropriateStyle(FormatterAssemblyStyle style, bool exceptionExpected)
+        {
+            var f = new BinaryFormatter();
+            var s = new MemoryStream();
+            f.Serialize(s, new Version1ClassWithoutField());
+            s.Position = 0;
+
+            f = new BinaryFormatter() { AssemblyFormat = style };
+            f.Binder = new DelegateBinder { BindToTypeDelegate = (_, __) => typeof(Version2ClassWithoutOptionalField) };
+            if (exceptionExpected)
+            {
+                Assert.Throws<SerializationException>(() => f.Deserialize(s));
+            }
+            else
+            {
+                var result = (Version2ClassWithoutOptionalField)f.Deserialize(s);
+                Assert.NotNull(result);
+                Assert.Equal(null, result.Value);
+            }
+        }
+
+        [Theory]
+        [InlineData(FormatterAssemblyStyle.Simple)]
+        [InlineData(FormatterAssemblyStyle.Full)]
+        public void OptionalField_Missing_Success(FormatterAssemblyStyle style)
+        {
+            var f = new BinaryFormatter();
+            var s = new MemoryStream();
+            f.Serialize(s, new Version1ClassWithoutField());
+            s.Position = 0;
+
+            f = new BinaryFormatter() { AssemblyFormat = style };
+            f.Binder = new DelegateBinder { BindToTypeDelegate = (_, __) => typeof(Version2ClassWithOptionalField) };
+            var result = (Version2ClassWithOptionalField)f.Deserialize(s);
+            Assert.NotNull(result);
+            Assert.Equal(null, result.Value);
+        }
+
+        public static IEnumerable<object[]> Deserialize_FuzzInput_MemberData()
+        {
+            var rand = new Random(42);
+            for (int i = 0; i < 100; i++)
+            {
+                byte[] data = new byte[rand.Next(100, 200)];
+                rand.NextBytes(data);
+                yield return new object[] { data };
+            }
```

Comment:
Good point. That kind of fuzzing then could be a separate test, the test would verify that either nothing happened, or there was a SerializationException. If there was eg a NullReferenceException, the test would fail.

---
In reply to: [71001854](https://github.com/dotnet/corefx/pull/10088#discussion_r71001854) [](ancestors = 71001854)